### PR TITLE
Fix kube_services.PublicAddr

### LINKED ISSUE #8101: The public_addr should be respected for kubernetes_service
### What

The kubernetes_service has a public_addr configuration option. This seems to be ignored. Instead, it seems to use the IP that the proxy_service sees the kubernetes_service as connecting from, and the bind port that the kubernetes_service is using.

public_addr should be used by the proxy_service when determining how to make the connection.

### How

When determining where the proxy_service should connect when accessing a kubernetes backend, use the public_addr as defined in the kubernetes_service config section.

### Why

This is especially important in networking environments where IOT is not ideal, and teleport simply needs to have the correct hostname/ip and port fed to it so it can make the connection

### Workaround

none

### LINKED ISSUE #8144: Kubernetes Access does not respect public_addr
Kubernetes Access does not respect `public_addr`. Suppose you configure a Kubernetes agent like the following:

```yaml
teleport:
  auth_servers:
    - "proxy.teleport-system.svc.cluster.local:3025"
  auth_token: "foo"
  data_dir: "/tmp/test_new_service_pod/data"
auth_service:
  enabled: no
proxy_service:
  enabled: no
ssh_service:
  enabled: no
kubernetes_service:
  enabled: "yes"
  kube_cluster_name: "kube1"
  public_addr: ["kube.teleport-system.svc.cluster.local:3027"]
  listen_addr: 0.0.0.0:3027
```

After a `tsh kube login` if you attempt to use `kubectl` you will see the following:

```
$ kubectl get pods
Error from server: Teleport proxy failed to connect to "kube" agent "kube.teleport-system.svc.cluster.local" over reverse tunnel:

  no tunnel connection found: no kube reverse tunnel for 8b1aecff-e64e-4edb-b936-088f4164d9e6.auth-proxy-5685cfccf4-chsnz found

This usually means that the agent is offline or has disconnected. Check the
agent logs and, if the issue persists, try restarting it or re-registering it
with the cluster.
```

This is because the `public_addr` field is not being set as can be seen in [lib/kube/proxy/server.go#L258-L281](https://github.com/gravitational/teleport/blob/master/lib/kube/proxy/server.go#L258-L281).

The following patch appears to fix the issue by simply propagating `public_addr` along. However, it needs to be cleaned up (don't index without checking length), test coverage needs to be added, and verified this is the expected behavior as described in [RFD 5 - Kubernetes (k8s) service enhancements](https://github.com/gravitational/teleport/blob/master/rfd/0005-kubernetes-service.md).

```diff
diff --git a/lib/kube/proxy/forwarder.go b/lib/kube/proxy/forwarder.go
index 585237eed..dd6140707 100644
--- a/lib/kube/proxy/forwarder.go
+++ b/lib/kube/proxy/forwarder.go
@@ -134,6 +134,8 @@ type ForwarderConfig struct {
    DynamicLabels *labels.Dynamic
    // LockWatcher is a lock watcher.
    LockWatcher *services.LockWatcher
+   // PublicAddr
+   PublicAddr string
 }

 // CheckAndSetDefaults checks and sets default values
diff --git a/lib/kube/proxy/server.go b/lib/kube/proxy/server.go
index 19a700e8c..86b6a1a97 100644
--- a/lib/kube/proxy/server.go
+++ b/lib/kube/proxy/server.go
@@ -256,7 +256,9 @@ func (t *TLSServer) GetServerInfo() (types.Resource, error) {
    defer t.mu.Unlock()

    var addr string
-   if t.listener != nil {
+   if t.TLSServerConfig.ForwarderConfig.PublicAddr != "" {
+       addr = t.TLSServerConfig.ForwarderConfig.PublicAddr
+   } else if t.listener != nil {
        addr = t.listener.Addr().String()
    }

diff --git a/lib/service/kubernetes.go b/lib/service/kubernetes.go
index 977a760dc..f6a40f531 100644
--- a/lib/service/kubernetes.go
+++ b/lib/service/kubernetes.go
@@ -237,6 +237,7 @@ func (process *TeleportProcess) initKubernetesService(log *logrus.Entry, conn *C
            StaticLabels:      cfg.Kube.StaticLabels,
            DynamicLabels:     dynLabels,
            LockWatcher:       lockWatcher,
+           PublicAddr:        cfg.Kube.PublicAddrs[0].String(),
        },
        TLS:           tlsConfig,
        AccessPoint:   accessPoint,
```

### PR DESCRIPTION
**Pull request by Joerger**
Closes https://github.com/gravitational/teleport/issues/8144
Closes https://github.com/gravitational/teleport/issues/8101

### CODE REVIEW

**Review by andrejtokarcik — COMMENTED** [review]
Should the version bumps really occur as part of this PR?

File: lib/kube/proxy/forwarder.go [hunk 1]
```diff
@@ -137,6 +137,8 @@ type ForwarderConfig struct {
 	// CheckImpersonationPermissions is an optional override of the default
 	// impersonation permissions check, for use in testing
 	CheckImpersonationPermissions ImpersonationPermissionsChecker
+	// PublicAddr is the addr that can be used to reach the kube cluster
```
Comment (zmb3) [inline lib/kube/proxy/forwarder.go:140]:
```suggestion
	// PublicAddr is the address that can be used to reach the kube cluster
```

File: lib/service/kubernetes.go [hunk 2]
```diff
@@ -218,6 +218,11 @@ func (process *TeleportProcess) initKubernetesService(log *logrus.Entry, conn *C
 		Streamer: streamer,
 	}
 
+	var publicAddr string
+	if len(cfg.Kube.PublicAddrs) > 0 {
```
Comment (zmb3) [inline lib/service/kubernetes.go:222]: Under what circumstances would someone configure multiple public addresses? Is there any special handling we need to do in that case?

File: Makefile [hunk 3]
```diff
@@ -11,7 +11,7 @@
 #   Stable releases:   "1.0.0"
 #   Pre-releases:      "1.0.0-alpha.1", "1.0.0-beta.2", "1.0.0-rc.3"
 #   Master/dev branch: "1.0.0-dev"
-VERSION=7.0.0-beta.1
```
Comment (zmb3) [inline Makefile:14]: Are all these version changes necessary here?

Comment (Joerger) [reply Makefile]: They're for cutting a dev build but aren't needed in the PR itself, will remove.

File: lib/service/kubernetes.go (hunk 2 above)
Comment (Joerger) [reply lib/service/kubernetes.go:222]: Providing more than one address doesn't seem to be really supported. I think this was done this way to incorporate the` utils.Strings` parser and match other configuration fields for UX.

Only the first address provided is used for dialing/heart-beating, but all provided addresses are used as principals for kube TLS certificates. @awly is this what was intended, or should `kube_service` be updated to try connecting over all provided `public_addr`'s?

Either way I think adding logic for this special case is out of scope for this bug fix.

**Review by tcsc — APPROVED** [review]

**Review by andrejtokarcik — APPROVED** [review]

**Review by zmb3 — APPROVED** [review]

**Review by klizhentas — APPROVED** [review]

**Review by tcsc — APPROVED** [review]

File: lib/service/kubernetes.go (hunk 2 above)
Comment (awly) [reply lib/service/kubernetes.go:222]: I agree it's out of scope.
Generally, we support multiple `public_addrs` in services for TLS cert principals - a user may have several addresses for a service (e.g. one external-facing and one for internal networks, with separate firewall rules) and TLS certs should validate against all of them.

For internal routing using just the first address is OK for now, I haven't heard of anyone complaining.

**Pull request merged**